#! /usr/bin/Rscript

#######################################################################################
### Run-c2g.R - V0.1 - Clement Goubert (2020) - goubert.clement@gmail.com           ###
### ------------------------------------------------------------------------------- ###
### This script uses the R function consensus2genome to infer breakpoints           ###
### to split and/or trim TE                                                         ###
### consensi generated by tools like RepeatModeler2, EDTA, REPET,...                ###
### USAGE:  TE-slocer.sh <TE-header> <[>                                            ###                            ###
### query: path to query (fasta file)                                               ###
### db: path to blast db (blast formated nucleotide database)                       ###
#######################################################################################



Args 		=	commandArgs(trailingOnly = TRUE)
query		=	as.character(Args[1])
database	=	as.character(Args[2])
evalue		=	as.numeric(Args[3])
FLthresh 	=	as.numeric(Args[4])
alpha 		=	as.numeric(Args[5])
full_alpha  =	as.numeric(Args[6])
autoy		=	as.character(Args[7]) # TRUE or numeric value for y max
output		=	as.character(Args[8])
selfdb		=	as.character(Args[9]) # bool
blastp      =   as.character(Args[10]) # includes orfs position
osize		=	as.numeric(Args[11])
wdir        =   as.character(Args[12]) # from the shell: path to running directory
title       =   as.character(Args[13])
v_x_line_1  =   as.character(Args[14])
v_x_line_2  =   as.character(Args[15])
tables      =   as.character(Args[16])

source(paste(wdir, "/", "consensus2genome.R", sep = ""))
source(paste(wdir, "/", "blastndotplot.R", sep = ""))

library(grid)

# Calculate the length of the sequence by calling the getlength.sh script
sequence_length <- as.numeric(system(paste(wdir,"/getlength.sh ", query, sep = ""), intern = TRUE))

# Set the default title
pdf_title <- paste("TE consensus", as.character(sequence_length), "bp")

# Check the value of tm and update the title accordingly
if (title == "after") {
  pdf_title <- paste("After TEtrimmer", as.character(sequence_length), "bp")
} else if (title == "extend") {
  pdf_title <- "After TEtrimmer Extended plot Blue lines are boundaries"
} else if (title == "before"){
  pdf_title <- paste("Before TEtrimmer", as.character(sequence_length), "bp")
}

pdf(
  width = 27,
  height = 7,
  file = paste(output, "/", tail(strsplit(as.character(query), "/")[[1]], 1), ".c2g.pdf", sep = "")
)
#pdf(width = 16, height = 16, file = paste(tail(strsplit(as.character(query), "/")[[1]],1), ".c2g.pdf", sep=""))

# Adjust margins to leave space for the title at the top
par(mar = c(6, 5, 4 + 4, 2) + 0.5)

# Set layout for a 2x2 plot
par(mfrow = c(1, 4))

print("R: ploting genome blastn results and computing coverage...")

consensus2genome(query 		=	query,
                 db 		=	database,
                 evalue 	=	evalue,
                 FL_thresh 	=	FLthresh,
                 alpha 		=	alpha,
                 full_alpha	=	full_alpha,
                 auto_y 	=	autoy,
                 bins       =   wdir,
                 output     =   output,
                 v_x_line_1 = v_x_line_1,
                 v_x_line_2 = v_x_line_2
#                 ,
#                 cover		=	cover,
#                 cov_thresh	=	drops
#                 ,
                 )

print("R: ploting self dot-plot and orf/protein hits...")

blastdotplot(query  =  query,
             db     =  selfdb,
             blast  =  blastp,
             os     =  osize,
             tables =  tables,
             output =  output)

title(main = pdf_title, outer = TRUE, line = -3, cex.main = 3)

dev.off()
